File: py/tree.py
MOD = (10**9) + 7

line = raw_input()
a = line.split(" ")
n = int(a[0])
k = int(a[1])

if(k==1):
	ans = n**(n-2)
elif(k==2):
	ans = (((2*n) - 2)**n)%MOD
	ans = ans*((2*n)**(n-2))
else:
	ans = (((3*n) - 3)**(2*n))%MOD
	ans = ans*((3*n)**(n-2))

ans = int(ans)%MOD
if(ans<0):
	ans = 0
print int(ans)

File: c,c++/tree.cpp
//Author: KewlJerk
#include <cstdio>
#include <iostream>

using namespace std;

#define MOD 1000000007
#define s(a) scanf("%lld",&a)
#define ll long long

ll int exp(ll int p, ll int q){
	if(q<0)
		return 0;

	ll int r = 1;
	while (q != 0) {
		if (q % 2 == 1) {
			r *= p;
			q--;

			if(r>=MOD)
				r %= MOD;
		}
		p *= p;
		q /= 2;

		if(p>=MOD)
			p %= MOD;

	}
	return r;
}

int main() {
	ll int n, k, ans;
	s(n), s(k);
	
	if(k==1){
		ans = exp(n, n-2);
	} else if(k==2){
		ans = exp(n, 2*(n-2));
		ans = (ans*exp(2, n-1))%MOD;
		ans = (ans*n)%MOD;
		ans = (ans*(n-1))%MOD;
	} else{
		ans = exp(n, 3*(n-2));
		ans = (ans*exp(6, n-1))%MOD;
		ans = (ans*n)%MOD;
		ans = (ans*(n-1))%MOD;
		ans = (ans*(n-1))%MOD;
	}

	printf("%lld\n", ans);

	return 0;
}